### PR TITLE
Lower MatrixNativeReader into manipulations of TableReads

### PR DESCRIPTION
**Pull request by tpoterba**
Add TableZipUnchecked IR node (used to zip rows and entries).

### CODE REVIEW

**Comment by tpoterba** [comment]
randomly assigned @catoverdrive

File: hail/src/main/scala/is/hail/variant/MatrixTable.scala [hunk 1]
```diff
@@ -127,14 +127,25 @@ abstract class AbstractMatrixTableSpec extends RelationalSpec {
...
   hail_version: String,
   references_rel_path: String,
   matrix_type: MatrixType,
-  components: Map[String, ComponentSpec]) extends AbstractMatrixTableSpec
+  components: Map[String, ComponentSpec]) extends AbstractMatrixTableSpec {
+
+  // some legacy files written as MatrixTableSpec wrote the wrong type to the entries table metadata
+  override def entriesTableSpec(path: String): AbstractTableSpec = {
+    val writtenETS = super.entriesTableSpec(path).asInstanceOf[TableSpec]
+    writtenETS.copy(table_type = TableType(matrix_type.entriesRVType, FastIndexedSeq(), matrix_type.globalType))
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/variant/MatrixTable.scala]: the entries table spec uses the MatrixTable global type? Aren't they stored separately?

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala [hunk 2]
```diff
@@ -868,7 +868,9 @@ object IRParser {
           val reader = try {
             Serialization.read[TableReader](readerStr)
           } catch {
-            case e: MappingException => throw e.cause
+            case e: MappingException =>
+              println(readerStr)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/Parser.scala]: remove/turn into a logging statement?

File: hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 3]
```diff
@@ -164,97 +169,66 @@ case class MatrixNativeReader(path: String) extends MatrixReader {
...
-            val rv3 = RegionValue(region)
-            new Iterator[RegionValue] {
-              def hasNext = {
-                val hn1 = it1.hasNext
-                val hn2 = it2.hasNext
-                assert(hn1 == hn2)
-                hn1
-              }
+      val entries: TableIR = TableRead(
+        TableType(
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala]: Can't these types get pulled directly from the spec, too? Should we make sure they match?

File: hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 4]
```diff
@@ -164,97 +169,66 @@ case class MatrixNativeReader(path: String) extends MatrixReader {
...
+    val entriesPath = path + "/entries"
+    val colsPath = path + "/cols"
+
     val hc = HailContext.get
 
-    val requestedType = mr.typ
-    assert(PruneDeadFields.isSupertype(requestedType, spec.matrix_type),
-      s"R: ${ requestedType.parsableString() }\nS: ${ spec.matrix_type.parsableString() }")
+    var tr: TableIR = TableRead(
+      TableType(
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala]: The globals are stored in their own file, right? I don't see them being read in anywhere? (Do we have tests for this?)

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: oopsie

File: hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala (hunk 4 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala]: (discussed, but for other prying eyes: the rows/cols/entries all have relative links to the globals and are read in with the globals)

File: hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala]: the isSupertype check should be sufficient

File: hail/src/main/scala/is/hail/variant/MatrixTable.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/variant/MatrixTable.scala]: as discussed, the globals are a relative path.

**Pull request merged**